eth/protocols/eth/broadcast.go

```diff
@@ -142,13 +142,17 @@ func (p *Peer) announceTransactions() {
 		if done == nil && len(queue) > 0 {
 			// Pile transaction hashes until we reach our allowed network limit
 			var (
-				count   int
-				pending []common.Hash
-				size    common.StorageSize
+				count        int
+				pending      []common.Hash
+				pendingTypes []byte
+				pendingSizes []uint32
+				size         common.StorageSize
 			)
 			for count = 0; count < len(queue) && size < maxTxPacketSize; count++ {
-				if p.txpool.Get(queue[count]) != nil {
+				if tx := p.txpool.Get(queue[count]); tx != nil {
 					pending = append(pending, queue[count])
+					pendingTypes = append(pendingTypes, tx.Type())
+					pendingSizes = append(pendingSizes, 1+uint32(tx.Size())) // txsize += 1 because of txtype byte
```

Comment:
This +1 seems bad to me. A legacy tx can also be sent untagged, won't that choke? IMHO we should not need to hack in +1 anywhere, otherwise its gonna blow up at some point where we forget.

Reply:
The `Size()` method shoud use `tx.MarshalBinary()` instead of `rlp.Encode(&c, &tx.inner)` in order for it to be correct, no?

Reply:
Actually, seems simpler to do `+1` unless legacy. The `encodeTyped` takes a concrete buffer, so requires alloc.

Reply:
We can just add more testcases, to ensure that `Size()` does equal the `MarshalBinary()` size all tx types

Reply:
The important thing is to make it non-ambiguous.

The tx.Size() returning only the inner size but no type prefix is a bit wonky imho. We currently only use it as a heuristic for broadcast limiting. That's kind of ok I guess, but if we want to use it as a precise protocol measure, we should have it include everything. I'm happy to extend the `Size` impl to do RLP encoding and +1 for non legacy. But this legacy +1 byte should not escape out into this PR, that concerns tx encofing, not network protocol.